Mathlib/MeasureTheory/Function/Holder.lean

```diff
@@ -0,0 +1,255 @@
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jireh Loreaux
+-/
+import Mathlib.MeasureTheory.Integral.Bochner
+
+/-! # Continuous bilinear maps on `MeasureTheory.Lp` spaces
+
+Given a continuous bilinear map `B : E →L[𝕜] F →L[𝕜] G`, we define an associated map
+`ContinuousLinearMap.holder : Lp E p μ → Lp F q μ → Lp G r μ` where `p q r` are a Hölder triple.
+We bundle this into a bilinear map `ContinuousLinearMap.holderₗ` and a continuous
+bilinear map `ContinuousLinearMap.holderL` under some additional assumptions.
+
+We also declare a heterogeneous scalar multiplication (`HSMul`) instance on `MeasureTheory.Lp`
+spaces. Although this could use the `ContinuousLinearMap.holder` construction above, we opt not to
+do so in order to minimize the necessary type class assumptions.
+
+When `p q : ℝ≥0∞` are Hölder conjugate (i.e., `HolderConjugate p q`), we also construct the
+natural map `ContinuousLinearMap.lpPairing : Lp E p μ →L[𝕜] Lp F q μ →L[𝕜] G` given by
+`fun f g ↦ ∫ x, B (f x) (g x) ∂μ`. When `B := (NormedSpace.inclusionInDoubleDual 𝕜 E).flip`, this
+is the natural map `Lp (Dual 𝕜 E) p μ →L[𝕜] Dual 𝕜 (Lp E q μ)`.
+ -/
+
+open ENNReal MeasureTheory Lp
+open scoped NNReal
+
+noncomputable section
+
+/-! ### Induced bilinear maps -/
+
+section Bilinear
+
+variable {α 𝕜 E F G : Type*} {m : MeasurableSpace α} {μ : Measure α}
+    {p q r : ENNReal} [hpqr : HolderTriple p q r] [NontriviallyNormedField 𝕜]
+    [NormedAddCommGroup E] [NormedAddCommGroup F] [NormedAddCommGroup G]
+    [NormedSpace 𝕜 E] [NormedSpace 𝕜 F] [NormedSpace 𝕜 G]
+    (B : E →L[𝕜] F →L[𝕜] G)
+
+namespace ContinuousLinearMap
+
+/-- The map between `MeasuryTheory.Lp` spaces satisfying `ENNReal.HolderTriple`
+induced by a continuous bilinear map on the underlying spaces. -/
+def holder (f : Lp E p μ) (g : Lp F q μ) : Lp G r μ :=
+  Memℒp.toLp (fun x ↦ B (f x) (g x)) <| by
+    refine .of_bilin (B · ·) ‖B‖₊ (Lp.memℒp f) (Lp.memℒp g) ?_ <|
+      .of_forall fun _ ↦ B.le_opNorm₂ _ _
+    exact B.aestronglyMeasurable_comp₂ (Lp.memℒp f).1 (Lp.memℒp g).1
+
+lemma coeFn_holder (f : Lp E p μ) (g : Lp F q μ) :
+    (B.holder f g : Lp G r μ) =ᵐ[μ] fun x ↦ B (f x) (g x) := by
+  rw [holder]
+  exact Memℒp.coeFn_toLp _
+
+lemma nnnorm_holder_apply_apply_le (f : Lp E p μ) (g : Lp F q μ) :
+    ‖(B.holder f g : Lp G r μ)‖₊ ≤ ‖B‖₊ * ‖f‖₊ * ‖g‖₊ := by
+  simp_rw [← ENNReal.coe_le_coe, ENNReal.coe_mul, ← enorm_eq_nnnorm, Lp.enorm_def]
+  apply eLpNorm_congr_ae (coeFn_holder B f g) |>.trans_le
+  exact eLpNorm_le_eLpNorm_mul_eLpNorm_of_nnnorm (Lp.memℒp f).1 (Lp.memℒp g).1 (B · ·) ‖B‖₊
+    (.of_forall fun _ ↦ B.le_opNorm₂ _ _)
+
+lemma norm_holder_apply_apply_le (f : Lp E p μ) (g : Lp F q μ) :
+    ‖(B.holder f g : Lp G r μ)‖ ≤ ‖B‖ * ‖f‖ * ‖g‖ :=
+  NNReal.coe_le_coe.mpr <| nnnorm_holder_apply_apply_le B f g
+
+lemma holder_add_left (f₁ f₂ : Lp E p μ) (g : Lp F q μ) :
+    (B.holder (f₁ + f₂) g : Lp G r μ) = B.holder f₁ g + B.holder f₂ g := by
```

Comment:
Should `r` be explicit in `holder` so as to avoid having to type-ascript `Lp G r μ` everywhere?